### ISSUE TITLE
Stopping domain takes a long time

### ISSUE THREAD

**Issue by heinrich-ulbricht** [QubesOS/qubes-issues#5426]
**Qubes OS version**
R4.0

**Affected component(s) or functionality**
Shutting down a domain.

**Brief summary**
Shutting down one of my domains takes ~40 seconds. Other domains only take ~6 seconds.

**To Reproduce**
Don't know how to reproduce. Works (not) on my machine:tm:.

**Expected behavior**
Domain shuts down fast.

**Actual behavior**
Domain takes ages.

**Additional context**
My domain has 600 GB private storage. I recently increased the size by ~60 GB, ignoring the advice to increase in steps of 10 GB. I just added 60 GB. Now I don't know if the slow shutdown is since then or if it is since I switched the template from Fedora 28 to Fedora 30 or not related at all. Somebody familiar with the inner workings might know if some of those steps could matter.

**Solutions you've tried**
Don't know where to start, which logs to consult or which tools to use.

Any help to get started diagnosing this is appreciated.

**Comment by marmarek** [comment]
Take a looks at `/var/log/xen/console/guest-VMNAME.log` (also available in qubes manager right click menu) for that domain. It should contain individual service stops during domain shutdown, including timestamps - it will give a hint what takes this long.

**Comment by heinrich-ulbricht** [comment]
@marmarek Looking at the most recent log file for the domain I can see that the log messages report successful shutdown almost immediately after clicking the "Shutdown" button in the Qubes menu. When I click Shutdown and look at the log I immediately see this:

```
[.[0;32m  OK  .[0m] Reached target .[0;1;39mShutdown.[0m.
[.[0;32m  OK  .[0m] Reached target .[0;1;39mFinal Step.[0m.
[.[0;32m  OK  .[0m] Started .[0;1;39mPower-Off.[0m.
[.[0;32m  OK  .[0m] Reached target .[0;1;39mPower-Off.[0m.
[   35.852900] systemd-shutdown[1]: Syncing filesystems and block devices.
[   35.915401] systemd-shutdown[1]: Sending SIGTERM to remaining processes...
[   35.922197] systemd-journald[210]: Received SIGTERM from PID 1 (systemd-shutdow).
[   35.968467] systemd-shutdown[1]: Sending SIGKILL to remaining processes...
[   35.976329] systemd-shutdown[1]: Unmounting file systems.
[   35.980770] [1619]: Remounting '/usr/lib/modules' read-only in with options 'lowerdir=/tmp/modules,upperdir=/sysroot/lib/modules,workdir=/sysroot/lib/.modules_work'.
[   35.985098] [1620]: Unmounting '/usr/lib/modules'.
[   35.995188] [1621]: Remounting '/' read-only in with options 'discard'.
[   36.017008] EXT4-fs (xvda3): re-mounted. Opts: discard
[   36.027040] systemd-shutdown[1]: All filesystems unmounted.
[   36.027075] systemd-shutdown[1]: Deactivating swaps.
[   36.027192] systemd-shutdown[1]: All swaps deactivated.
[   36.027219] systemd-shutdown[1]: Detaching loop devices.
[   36.027512] systemd-shutdown[1]: All loop devices detached.
[   36.027542] systemd-shutdown[1]: Detaching DM devices.
[   36.172314] reboot: System halted
```

And then it takes ~40 seconds. No further log entries are added, so the log does not really help. Is this ok? Where would I look further?

**Comment by brendanhoar** [comment]
1. What physical storage are you using, SSD or HDD?
2. Are you configured with the standard LVM-thin, btrfs or standard file-based containers?
3. What was the initial version of Qubes installed?
4. Are you running standard or testing repos?

**Comment by heinrich-ulbricht** [comment]
@brendanhoar 

1. classic HDD
1. standard LVM-thin in a separate pool
1. Qubes R4.0
    * did a migration lately, but from 4.0 to 4.0 to increase the size of my main-pool SSD; re-created the large pool on another HDD that SHOULD be the same model as the first one
    * I upgraded the underlying Fedora 28 TemplateVM to Fedora 30 lately
1. standard Qubes and Fedora repos

Your questions lead me to do some more testing:
* I switched the TemplateVM to Fedora 30 and Fedora 28 Minimal - no change, the long shutdown time persists

Shutdown time currently is ~50 seconds in all cases.

The first time I noticed the long shutdown is after increasing the private storage by a large amount. But it might be that I did not notice the long shutdown time before during regular shutdown. But I'm certain it used to be a lot faster. If I could pinpoint it to a slow HDD or something this would be fine. I'd just like to see some log where the delay can be seen. Maybe it's a timeout or something.

**Comment by heinrich-ulbricht** [comment]
Just had look into the oldest xen log file I have, which is from **Oct 06**. It looks different for shutdown:
```
         Starting Power-Off...
[25187.655862] systemd-shutdow: 25 output lines suppressed due to ratelimiting
[25187.759906] systemd-shutdown[1]: Syncing filesystems and block devices.
[25187.839439] systemd-shutdown[1]: Sending SIGTERM to remaining processes...
[25187.842755] systemd-journald[203]: Received SIGTERM from PID 1 (systemd-shutdow).
[25187.874552] systemd-shutdown[1]: Sending SIGKILL to remaining processes...
[25187.877111] systemd-shutdown[1]: Unmounting file systems.
[25187.879350] [2493]: Remounting '/usr/lib/modules' read-only in with options 'lowerdir=/tmp/modules,upperdir=/sysroot/lib/modules,workdir=/sysroot/lib/.modules_work'.
[25187.887041] [2494]: Unmounting '/usr/lib/modules'.
[25187.896060] [2495]: Remounting '/' read-only in with options 'discard'.
[25187.923818] EXT4-fs (xvda3): re-mounted. Opts: discard
[25187.927815] systemd-shutdown[1]: All filesystems unmounted.
[25187.927830] systemd-shutdown[1]: Deactivating swaps.
[25187.927869] systemd-shutdown[1]: All swaps deactivated.
[25187.927880] systemd-shutdown[1]: Detaching loop devices.
[25187.928048] systemd-shutdown[1]: All loop devices detached.
[25187.928060] systemd-shutdown[1]: Detaching DM devices.
[25187.954533] xenbus: xenbus_dev_shutdown: device/vbd/51840: Closed != Connected, skipping
[25187.993300] reboot: System halted
```
The _xenbus_ line is gone in the current log. But then again those changes might be to kernel or xen updates. And the delay is not covered in the log.

**Comment by heinrich-ulbricht** [comment]
guid.domain.log differs from other domains. It shows:
```
Icon size: 128x128
libvchan_is_eof
Icon size: 128x128
domain dead
Failed to connect to gui-agent
```
These log entries seem to be unique to this domain:
```
domain dead
Failed to connect to gui-agent
```
Maybe it's just a reporting issue? Somebody not noticing the shutdown already happened?

**Comment by marmarek** [comment]
Those particular messages are probably unrelated, but I have another idea: during that long shutdown observe output of `xl list`, specifically state column. Is it `-b----` or something else when that long delay happens?
If something else, check what version of `qubes-libvchan-xen` do you have, specifically if 4.0.7 or older - a related change was included there (that could improve this).

**Comment by heinrich-ulbricht** [comment]
@marmarek 
During shutdown I can observe the `state` in `xl list` change like this over time (_domain_ being the name of my domain):
* _domain_: r----- (time: 14.1)
* _domain_: -b---- (time: 20.2)
* _domain_: r----- (21.0)
* _(null)_: --ps-- (21.0)
* _(null)_: --ps-d (21.0)

Note how the domain name at some point is replaced by _(null)_ and the time stops.

It stays like _(null): --ps-d (21.0)_ for about 8 seconds, then the entry disappears while the domain is still shutting down for many more seconds according to the Qubes Domains list.

In my AppVM I have version 4.0.7 of _qubes-libvchan-xen_.

```
[user@domain ~]$ dnf list installed | grep qubes-libvchan-xen
qubes-libvchan-xen.x86_64                   4.0.7-1.fc30                          @qubes-vm-r4.0-current
```

**Comment by brendanhoar** [comment]
What model of HDD is this? Seagate? Does it report it supports ... trim?

**Comment by heinrich-ulbricht** [comment]
@brendanhoar 
1. It's a Western Digital WD10EAVS from years ago. So SHR, trim etc. shouldn't be things to consider (right?)
1. Yes, I _temporarily_ changed the qubes-wide default pool from the primary (SSD) pool to the secondary pool to restore the large domain to the HDD pool. Then I changed it back to the SSD pool.
1. The template is located on the SSD.
1. Find the grep of the running VM below; _pool00_ is on the SSD, _vg_contoso_/_lv_contoso_ on the HDD

```
[heinrich@dom0 console]$ sudo lvs | grep contoso
  vm-contoso-root-snap                               qubes_dom0 Vwi-aotz--  21.72g pool00    vm-fedora-30                         56.84                                  
  lv_contoso                                         vg_contoso  twi-aotz-- 931.27g                                                             61.82  39.53                           
  vm-contoso-private                                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572293911-back                 98.26                                  
  vm-contoso-private-1572268964-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso                                                   98.26                                  
  vm-contoso-private-1572293911-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572268964-back                 98.26                                  
  vm-contoso-private-snap                            vg_contoso  Vwi-aotz-- 585.94g lv_contoso vm-contoso-private                                 98.26                                  
  vm-contoso-volatile                                vg_contoso  Vwi-aotz--  10.00g lv_contoso 
```

Do you see anything in there that looks odd?

**Comment by brendanhoar** [comment]
Thanks. 

I had a (now seemingly incorrect) hunch the issue was that blkdiscard was issuing discards* against thin-volumes hosted on the HDD...but that WD drive is around 10 years old, so it definitely does not support TRIM, so that's not the issue.

-Brendan

* Qubes issues blkdiscard before lvremove post-shutdown for volatile/snapshot volumes and/or post-VM-delete for deleted VMs. It's good practice, for several reasons, upper layers to notify lower layers that storage is being de-allocated, and lvm does not currently explicitly do so when lvremove is executed. If all layers have pass-down enabled (by default, Qubes doesn't, but many of us enable it), this leads to TRIM requests being issued to the physical storage layer.

**Comment by heinrich-ulbricht** [comment]
@brendanhoar By explicitly stating the age of this drive you make me wanna do backups more often...
Thanks for the theory and information. The mystery continues.

**Comment by heinrich-ulbricht** [comment]
I made another observation today while trying to reduce the number of `revisions_to_keep` for the pool to see if this changes something. (Not sure if this works, I'm still trying to make sense of #3256)

Calling `sudo lvs | grep contoso` repeatedly while the large domain is shutting down shows that the value in the _Data%_ column decreases steadily until it reaches zero. The "domain has shut down" message pops up in this exact moment when the value reaches zero.

Here is the log of calling `sudo lvs | grep contoso` three times (_Data%_ being 37.36, then 15.23, then 0.01). By the way, there is a noticeable delay of about ~3 seconds until each of those calls returns.

```
[heinrich@dom0 ~]$ sudo lvs | grep contoso
  lv_contoso                                         vg_contoso  twi-aotz-- 931.27g                                                             61.82  55.72                           
  vm-contoso-private                                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572532134-back                 98.26                                  
  vm-contoso-private-1572522261-back                 vg_contoso  Vwi-aotz-- 585.94g lv_contoso                                                   37.36                                  
  vm-contoso-private-1572532035-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572522261-back                 98.26                                  
  vm-contoso-private-1572532134-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572532035-back                 98.26                                  
[heinrich@dom0 ~]$ sudo lvs | grep contoso
  lv_contoso                                         vg_contoso  twi-aotz-- 931.27g                                                             61.82  62.31                           
  vm-contoso-private                                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572532134-back                 98.26                                  
  vm-contoso-private-1572522261-back                 vg_contoso  Vwi-aotz-- 585.94g lv_contoso                                                   15.23                                  
  vm-contoso-private-1572532035-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572522261-back                 98.26                                  
  vm-contoso-private-1572532134-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572532035-back                 98.26                                  
[heinrich@dom0 ~]$ sudo lvs | grep contoso
  lv_contoso                                         vg_contoso  twi-aotz-- 931.27g                                                             61.82  64.77                           
  vm-contoso-private                                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572532134-back                 98.26                                  
  vm-contoso-private-1572522261-back                 vg_contoso  Vwi-aotz-- 585.94g lv_contoso                                                   0.01                                   
  vm-contoso-private-1572532035-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572522261-back                 98.26                                  
  vm-contoso-private-1572532134-back                 vg_contoso  Vwi-a-tz-- 585.94g lv_contoso vm-contoso-private-1572532035-back                 98.26                                  
```

@marmarek @brendanhoar How would you interpret this? What exactly is happening there?

**Comment by marmarek** [comment]
That makes sense. Issuing DISCARD for the whole 580GB volume indeed can take time. LVM internally handle it too, even if underlying block doesn't support it. Specifically - release blocks from the volume and update internal metadata about those blocks (reference counters etc).
I see three options how to proceed here:
1. Don't do anything and accept long shutdown - I think the only things that are slowed down during this time are about the very some qube (for example if you want to start it again). Other qubes should be unaffected.
2. Don't issue DISCARD if underlying device doesn't support it. Issuing DISCARD here is mostly useful for the underlying SSD disk, not so much for HDD. On the other hand, detecting the support may be tricky (there are multiple layers to consider, including LUKS, LVM itself etc)
3. Make it more asynchronous, in the background - don't hold further operations of the qube while it's discarding old volume. That volume wouldn't be used in any subsequent operations anyway, so it should be doable, but introducing more concurrency into the process may be risky - for example we'd need to ensure the operation is completed when shutting down the system, or removing the qube.

Do you have a specific use case against assumptions in the first point?

**Comment by brendanhoar** [comment]
Some interplay of HDD, /etc/lvm/lvm.conf (and cryptsetup options, if being used) may make it possible that the blkdiscard issued before lvremove might be *zeroing* out the container files (instead of issuing discards, since the backing store is not trim-capable).

1. What is the setting of issue_discards= in /etc/lvm/lvm.conf?
2. What is the setting of thin_pool_discards in /etc/lvm/lvm.conf?
3. If cryptsetup is being used in an intervening layer between the HDD and the PV/VG/thin volume, is the --allow-discards option being used with the cryptsetup?

Lastly, I wonder if blkdiscard automatically falls back to using zeroout if the volume (real or virtual) does not indicate trim support. That would not be optimal behavior on a thin volume. I may have to go dig out some spinning rust from my closet to do some tests.

Brendan

**Comment by heinrich-ulbricht** [comment]
@brendanhoar 
1. `issue_discards` in /etc/lvm/lvm.conf is **0**
1. `thin_pool_discards` in /etc/lvm/lvm.conf is **commented out**, thus no explicit value set
1. Where can I check this? I set up the second pool including cryptsetup by following these instructions: https://www.qubes-os.org/doc/secondary-storage/ - no mentioning of discard options there, did configure nothing else

(Just as a side note, and this could be irrelevant: before I can start the (long-shutdown) AppVM in the second pool I have to start another VM (from the SSD pool) first, OR I have to call `sudo service qubesd restart`. Otherwise I get an error that the pool could not be found (have to reboot to get the exact error message). I think I commented somewhere else a long time ago about this but since then ignored it as this workaround was ok for me. Just wanted to mention this as we are taking deeper looks into the specifics of this pool.)

**Comment by heinrich-ulbricht** [comment]
@marmarek 
First and foremost I want to know if something is broken. If it is broken I want to repair it, if possible. 
If it is not broken then I want to know the options and the effort it takes (and for whom) to implement those. If it's too much hassle I can live with it.
I'm currently on a dual-boot machine and unfortunately have to switch between operating systems quite often. So every time I shut down Qubes I have this minute-long wait :-/ It's just annoying.

**Comment by marmarek** [comment]
> First and foremost I want to know if something is broken.

I wouldn't call it broken. Rather sub-optimal performance.

> So every time I shut down Qubes I have this minute-long wait :-/

This rules out my third idea, as you'd still need to wait during shutdown.

> What is the setting of issue_discards= in /etc/lvm/lvm.conf?

I believe this setting applies to non-thin volumes only.

> What is the setting of thin_pool_discards in /etc/lvm/lvm.conf?

This indeed is relevant for thin volumes, possible values: `ignore`, `nopassdown`, `passdown`. If lower layer doesn't support discard, `nopassdown` and `passdown` are effectively equivalent. Default is `passdown`.

>     3\. If cryptsetup is being used in an intervening layer between the HDD and the PV/VG/thin volume, is the --allow-discards option being used with the cryptsetup?

See https://www.qubes-os.org/doc/disk-trim/#luks

**Comment by iamahuman** [comment]
While shutting down a VM, I observe freezing in other VMs on the same pool as well. During the freeze, A `kworker/**-dm-thin` kernel thread takes 100% of CPU. I suppose this is the same issue?

**Comment by noskb** [comment]
I'm using an SSD but I had the same problem and it was only occurred on SanDisk SDSSDXPS480G and not on Samsung SSD 860 EVO.
Of course my domain has a large size of private storage.

In conclusion, it was solved by changing the discard mode of LVM thin pool from passdown to nopassdown.

`sudo lvchange --discards nopassdown qubes_dom0/pool00` in dom0 terminal

This will restore the physical space to the thin pool as in the past, but no TRIM command will be issued to the physical device.

[from the lvmthin manpage ](https://man7.org/linux/man-pages/man7/lvmthin.7.html)

    Discard

       The discard behavior of a thin pool LV determines how discard
       requests are handled.  Enabling discard under a file system may
       adversely affect the file system performance (see the section on
       fstrim for an alternative.)  Possible discard behaviors:

       ignore: Ignore any discards that are received.

       nopassdown: Process any discards in the thin pool itself and allow
       the no longer needed extents to be overwritten by new data.

       passdown: Process discards in the thin pool (as with nopassdown), and
       pass the discards down the the underlying device.  This is the
       default mode.




The default "passdown" may cause performance degradation in some cases.

**Comment by marmarek** [comment]
This sounds like your SSD driver has very poor implementation of "TRIM" operation. "nopassdown" indeed may improve performance in this case, but depending on SSD model, may reduce the disk life time. It is also possible that this disk firmware is stupid enough it won't matter at all besides performance. Hard to tell without doing some extensive tests or analyzing disk firmware.

**Comment by brendanhoar** [comment]
At a minimum, Qubes should document possible performance issues under LVM on VM shutdown (during volatile volume/snapshot removals) with both SMR HDs generally and SSDs with poor firmware. Docs should recommend “nopassdown “ in lvm.conf as the workaround, noting the caveats for SSDs above.

I was going to recommend exposing this option via the qubes global settings but thinking it through, with all the custom storage layouts out there it might get messy, so that’s probably not a great idea.

[I say this all as the instigator (or at least one of them) of adding the blkdiscard invocation to lvm volume removals. I still think the benefits outweigh the other issues.]

B

**Comment by andrewdavidwong** [comment]
> At a minimum, Qubes should document possible performance issues under LVM on VM shutdown (during volatile volume/snapshot removals) with both SMR HDs generally and SSDs with poor firmware. Docs should recommend “nopassdown “ in lvm.conf as the workaround, noting the caveats for SSDs above.

In case you (or any others reading this) are not already aware, the documentation is a community effort, and everyone is welcome to contribute. (That's how things like this get updated!) So, if you'd like to get involved with the project, this is a great way to do it. You can read more about how to submit documentation changes here:

https://www.qubes-os.org/doc/doc-guidelines/

**Comment by iamahuman** [comment]
dm-thin deletes the mapping nodes one at a time, and ftrace tells me the kernel is spending all the time rebalancing the btree and locking/unlocking nodes.

It certainly needs to get smarter and just delete the common ancestors--which in turn would automatically decrement refcounts (in spacemap?) of all the blocks allocated for child nodes via callbacks, if I read the kernel code correctly.

The most annoying thing about this bug is the total I/O freeze of other volumes in the same pool, resulting in other VMs potentially ceasing to respond for a few seconds.

**Comment by iamahuman** [comment]
After reading the device-mapper implementation myself and as per various online sources, I believe the solution (not workaround) is to use Btrfs or XFS, which has been around for 10+ years.

Or patch dm-thin, but I think it might be a waste of time...

**Comment by iamahuman** [comment]
Related: #5077
Introduced in: QubesOS/qubes-core-admin#267 (QubesOS/qubes-core-admin@62e3d571206f134adf6989345962fd92f103e2f3)

**Comment by iamahuman** [comment]
How about performing `blkdiscard` asynchronously? Possible solution would be to rename the LV to be removed to `vm-*-stale`, put it to reclaim queue, do blkdiscard, and finally run lvremove on it.

**Comment by iamahuman** [comment]
#6476 will resolve this.

**Comment by DemiMarie** [comment]
> After reading the device-mapper implementation myself, plus per various online sources, I believe the solution (not workaround) is to just use Btrfs or XFS. Both have been around for 10+ years.

Agreed, although I am wondering why you say this.

> Or patch dm-thin, but I think it might be a waste of time...

Why would it be a waste of time?

**Comment by iamahuman** [comment]
> > After reading the device-mapper implementation myself, plus per various online sources, I believe the solution (not workaround) is to just use Btrfs or XFS. Both have been around for 10+ years.
> 
> Agreed, although I am wondering why you say this.

Generally, the apparant lack of interest from maintainers. I might have mistaken, though.

> 
> > Or patch dm-thin, but I think it might be a waste of time...
> 
> Why would it be a waste of time?

Perhaps not completely worthless; I merely meant that It might not be worth the time in face of an alternative solution.

**Comment by DemiMarie** [comment]
> > > After reading the device-mapper implementation myself, plus per various online sources, I believe the solution (not workaround) is to just use Btrfs or XFS. Both have been around for 10+ years.
> > 
> > 
> > Agreed, although I am wondering why you say this.
> 
> Generally, the apparant lack of interest from maintainers. I might have mistaken, though.

In what areas?  dm-thin is a key part of LVM2, which is at the core of Red Hat’s storage offerings.  But I strongly suspect that their needs and ours do not match up very well, especially when it comes to startup and shutdown performance.

**Comment by iamahuman** [comment]
> > > > After reading the device-mapper implementation myself, plus per various online sources, I believe the solution (not workaround) is to just use Btrfs or XFS. Both have been around for 10+ years.
> > > 
> > > 
> > > Agreed, although I am wondering why you say this.
> > 
> > 
> > Generally, the apparant lack of interest from maintainers. I might have mistaken, though.
> 
> In what areas? dm-thin is a key part of LVM2, which is at the core of Red Hat’s storage offerings. But I strongly suspect that their needs and ours do not match up very well, especially when it comes to startup and shutdown performance.

Yes, that part exactly. IIRC this issue has been brought up in dm-devel ML a few times, with no response.

**Comment by DemiMarie** [comment]
> > > > > After reading the device-mapper implementation myself, plus per various online sources, I believe the solution (not workaround) is to just use Btrfs or XFS. Both have been around for 10+ years.
> > > > 
> > > > 
> > > > Agreed, although I am wondering why you say this.
> > > 
> > > 
> > > Generally, the apparant lack of interest from maintainers. I might have mistaken, though.
> > 
> > 
> > In what areas? dm-thin is a key part of LVM2, which is at the core of Red Hat’s storage offerings. But I strongly suspect that their needs and ours do not match up very well, especially when it comes to startup and shutdown performance.
> 
> Yes, that part exactly. IIRC this issue has been brought up in dm-devel ML a few times, with no response.

I wonder how much of the problem is dm-thin and how much is LVM2.  LVM2’s metadata format is textual and particularly inefficient.  We could probably do much better if we used dm-thin directly.

**Comment by DemiMarie** [comment]
> Yes, that part exactly. IIRC this issue has been brought up in dm-devel ML a few times, with no response.

Do you have any links?

**Comment by DemiMarie** [comment]
On further thought: according to https://lore.kernel.org/linux-lvm/540a8738-e09d-b80d-ba1b-bd91b3d6c3d5@redhat.com/ LVM will automatically perform the equivalent of `blkdiscard` if `issue_discards = 1` is set in `lvm.conf`.

**Comment by marmarek** [comment]
>  LVM2’s metadata format is textual and particularly inefficient.

There is `lvmetad` daemon that caches the parsed metadata, though. I haven't analyzed how good it performs, but I believe it should avoid most performance issues caused by loading metadata each time (whatever the format). My understanding of very basic performance analysis says the most performance impact is rather in a) various locking and b) synchronization with udev. There are options to configure or even bypass either of them, but I haven't came to any solution that is both faster and still safe.

**Comment by iamahuman** [comment]
> I wonder how much of the problem is dm-thin

See https://github.com/QubesOS/qubes-issues/issues/5426#issuecomment-761595524.

> and how much is LVM2.

Not very much relevant to performance here, I guess.

> LVM2’s metadata format is textual and particularly inefficient.

You mean the XML format used by `thin_dump` / `thin_restore`?

> We could probably do much better if we used dm-thin directly.

The actual sector mappings are stored as B-tree in the meta LV, which the `lvm` tool doesn't really care much.

**Comment by marmarek** [comment]
> LVM will automatically perform the equivalent of `blkdiscard` if `issue_discards = 1` is set in `lvm.conf`.

I believe it affects only fully provisioned LVs, not thin provisioned ones. But let me double check.

**Comment by DemiMarie** [comment]
> > LVM2’s metadata format is textual and particularly inefficient.
> 
> There is `lvmetad` daemon that caches the parsed metadata, though. I haven't analyzed how good it performs, but I believe it should avoid most performance issues caused by loading metadata each time (whatever the format). My understanding of very basic performance analysis says the most performance impact is rather in a) various locking and b) synchronization with udev. There are options to configure or even bypass either of them, but I haven't came to any solution that is both faster and still safe.

Can the locking be removed if we wrap a global lock around all LVM calls?  Also we can use the LVM shell (with JSON output).

**Comment by marmarek** [comment]
> I believe it affects only fully provisioned LVs, not thin provisioned ones. But let me double check.

Yup, it is the case.

<details>

```shell_session
bash-5.0# truncate -s 500M pv.img
bash-5.0# losetup /dev/loop0 ./pv.img
bash-5.0# vgcreate vg0 /dev/loop0
  Physical volume "/dev/loop0" successfully created.
  Volume group "vg0" successfully created
bash-5.0# lvcreate --type thin-pool -n pool0 -l 100%FREE vg0
  Thin pool volume with chunk size 64.00 KiB can address at most 15.81 TiB of data.
  Logical volume "pool0" created.
bash-5.0# lvcreate --thin-pool=pool0 -V 300M -n vol1 vg0
  Logical volume "vol1" created.
bash-5.0# ls -ls pv.img 
8320 -rw-r--r-- 1 root root 524288000 Mar 31 17:10 pv.img
bash-5.0# cat /dev/zero > /dev/vg0/vol1 
cat: write error: No space left on device
bash-5.0# ls -lsh pv.img 
309M -rw-r--r-- 1 root root 500M Mar 31 17:11 pv.img
bash-5.0# lvs
  LV    VG  Attr       LSize   Pool  Origin Data%  Meta%  Move Log Cpy%Sync Convert
  pool0 vg0 twi-aotz-- 488.00m              61.48  14.65                           
  vol1  vg0 Vwi-a-tz-- 300.00m pool0        100.00                                 
bash-5.0# blkdiscard /dev/vg0/vol1
bash-5.0# ls -lsh pv.img 
8.2M -rw-r--r-- 1 root root 500M Mar 31 17:11 pv.img
bash-5.0# cat /dev/zero > /dev/vg0/vol1 
cat: write error: No space left on device
bash-5.0# ls -lsh pv.img 
309M -rw-r--r-- 1 root root 500M Mar 31 17:11 pv.img
bash-5.0# lvmconfig devices/issue_discards
issue_discards=1
bash-5.0# ls -lsh pv.img 
309M -rw-r--r-- 1 root root 500M Mar 31 17:12 pv.img
bash-5.0# lvremove vg0/vol1
Do you really want to remove and DISCARD active logical volume vg0/vol1? [y/n]: y
  Logical volume "vol1" successfully removed
bash-5.0# ls -lsh pv.img 
309M -rw-r--r-- 1 root root 500M Mar 31 17:13 pv.img
```

**Comment by brendanhoar** [comment]
Due to lvm_thin not issuing discards down the device chain when removing a volume/snapshot, blkdiscard was added explicitly during the volume removal process during the qubes 4.0 lifecycle.

Is the performance issue related to this explicit blkdiscard?

If so, perhaps a global (or pool? or VM?) flag could be supported for those who want to disable blkdiscard for various reasons.

B

**Comment by DemiMarie** [comment]
> dm-thin deletes the mapping nodes one at a time, and ftrace tells me the kernel is spending all the time rebalancing the btree and locking/unlocking nodes.
> 
> It certainly needs to get smarter and just delete the common ancestors--which in turn would automatically decrement refcounts (in spacemap?) of all the blocks allocated for child nodes via callbacks, if I read the kernel code correctly.
> 
> The most annoying thing about this bug is the total I/O freeze of other volumes in the same pool, resulting in other VMs potentially ceasing to respond for a few seconds.

I wonder how hard a kernel patch would be.

**Comment by iamahuman** [comment]
> > Yes, that part exactly. IIRC this issue has been brought up in dm-devel ML a few times, with no response.
> 
> Do you have any links?

Turns out my memory served me wrong, although I'm not sure where I've got such impression. My apologies.

That said, here's some links I've been exploring:

- https://lore.kernel.org/dm-devel/35cf1ecf-5cd8-604b-ec4e-18c9fd4a4195@linux.alibaba.com/T/
- https://lore.kernel.org/dm-devel/4F9FDCC2.2080905@shiftmail.org/T/
- https://lore.kernel.org/linux-lvm/cc1e4d47824ca32d2e055fcb05473242@assyoma.it/T/

**Comment by iamahuman** [comment]
> > dm-thin deletes the mapping nodes one at a time, and ftrace tells me the kernel is spending all the time rebalancing the btree and locking/unlocking nodes.
> > It certainly needs to get smarter and just delete the common ancestors--which in turn would automatically decrement refcounts (in spacemap?) of all the blocks allocated for child nodes via callbacks, if I read the kernel code correctly.
> > The most annoying thing about this bug is the total I/O freeze of other volumes in the same pool, resulting in other VMs potentially ceasing to respond for a few seconds.
> 
> I wonder how hard a kernel patch would be.

One way that comes to my mind is to locate the "dominators" of all the leaf nodes in the range, and call `dm_btree_del` on them. This way, we can save much of re-balancing work while preserving refcount book-keeping.

Another issue that bothers me is that dm-thin performs discard in a "best effort" basis i.e. whenever there is OOM or heavy contention, it may abandon certain discard BIOs without remembering to do them later. It also won't attempt to retry discard operations that had been aborted due to e.g. a reboot or crash. [BTRFS is apparently more reliable in this respect.](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/btrfs/discard.c)

**Comment by rustybird** [comment]
To the people who experience these shutdown delays: If you shut down many VMs in parallel with `qvm-shutdown --wait vm1 vm2 vm3...`, is the total delay the same as the sum of the individual delays, less, or more?

**Comment by DemiMarie** [comment]
> > > dm-thin deletes the mapping nodes one at a time, and ftrace tells me the kernel is spending all the time rebalancing the btree and locking/unlocking nodes.
> > > It certainly needs to get smarter and just delete the common ancestors--which in turn would automatically decrement refcounts (in spacemap?) of all the blocks allocated for child nodes via callbacks, if I read the kernel code correctly.
> > > The most annoying thing about this bug is the total I/O freeze of other volumes in the same pool, resulting in other VMs potentially ceasing to respond for a few seconds.
> > 
> > 
> > I wonder how hard a kernel patch would be.
> 
> One way that comes to my mind is to locate the "dominators" of all the leaf nodes in the range, and call `dm_btree_del` on them. This way, we can save much of re-balancing work while preserving refcount book-keeping.

Would you be interested in creating a kernel patch for this?

it is also possible (using `thin_trim`) to perform the equivalent of `fstrim` for a thin pool.  This can only be done when the pool is not in use, but QubesOS could do it from the initramfs.

**Comment by ejose19** [comment]
In case this is useful, following @iamahuman https://github.com/QubesOS/qubes-issues/issues/3639#issuecomment-787459236, I've inspected what is causing near 100% single-core usage when shutting down the problematic VM, and here's the output of `sudo perf record -g -a -- qvm-shutdown --wait slow-vm`:

```
Samples: 2M of event 'cpu-clock', Event count (approx.): 675294250000
  Children      Self  Command          Shared Object                                     Symbol
+   91.85%    91.84%  swapper          [kernel.kallsyms]                                 [k] xen_hypercall_sched_op
+    5.93%     0.00%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] ret_from_fork
+    5.93%     0.00%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] kthread
+    5.93%     0.00%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] worker_thread
+    5.93%     0.00%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] process_one_work
+    5.93%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] do_worker
+    5.15%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] process_discard_bio
+    5.15%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] process_discard_cell_passdown
+    4.89%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] process_prepared_discard_passdown_pt1
+    3.87%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] dm_thin_remove_range
+    3.72%     0.02%  kworker/u24:8-k  [dm_persistent_data]                              [k] dm_btree_remove_leaves
+    3.69%     1.17%  kworker/u24:8-k  [dm_persistent_data]                              [k] dm_tm_shadow_block
+    3.18%     0.03%  kworker/u24:8-k  [dm_persistent_data]                              [k] sm_ll_mutate
+    2.14%     0.02%  kworker/u24:8-k  [dm_persistent_data]                              [k] shadow_step
+    2.08%     0.02%  kworker/u24:8-k  [dm_persistent_data]                              [k] rebalance_children
+    1.93%     0.09%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] new_read
+    1.85%     0.02%  kworker/u24:8-k  [dm_persistent_data]                              [k] init_child.isra.4
+    1.81%     0.03%  kworker/u24:8-k  [dm_persistent_data]                              [k] insert
+    1.71%     0.01%  kworker/u24:8-k  [dm_persistent_data]                              [k] disk_ll_save_ie
+    1.61%     0.01%  kworker/u24:8-k  [dm_persistent_data]                              [k] sm_disk_inc_block
+    1.58%     0.08%  kworker/u24:8-k  [dm_persistent_data]                              [k] btree_insert_raw
+    1.57%     0.10%  kworker/u24:8-k  [dm_persistent_data]                              [k] dm_bm_read_lock
+    1.54%     0.01%  kworker/u24:8-k  [dm_persistent_data]                              [k] sm_disk_dec_block
+    1.50%     0.06%  kworker/u24:8-k  [dm_persistent_data]                              [k] sm_ll_lookup_bitmap
+    1.46%     1.23%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] __bufio_new
+    1.40%     0.09%  kworker/u24:8-k  [dm_persistent_data]                              [k] sm_metadata_count_is_more_than_one
+    1.12%     0.02%  kworker/u24:8-k  [dm_persistent_data]                              [k] dm_btree_lookup
+    1.00%     0.04%  kworker/u24:8-k  [dm_persistent_data]                              [k] dm_bm_write_lock
+    0.94%     0.05%  kworker/u24:8-k  [dm_persistent_data]                              [k] btree_lookup_raw.constprop.7
+    0.82%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] dm_pool_inc_data_range
+    0.79%     0.00%  kworker/u24:8-k  [dm_persistent_data]                              [k] inc_children
+    0.76%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] process_prepared
+    0.71%     0.01%  kworker/u24:8-k  [dm_persistent_data]                              [k] ro_step
+    0.67%     0.01%  kworker/u24:8-k  [dm_persistent_data]                              [k] disk_ll_load_ie
+    0.64%     0.00%  kworker/u24:8-k  [dm_persistent_data]                              [k] rebalance2
+    0.63%     0.02%  kworker/u24:8-k  [dm_persistent_data]                              [k] dm_bm_unlock
+    0.54%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] process_prepared_discard_passdown_pt2
+    0.54%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] dm_pool_dec_data_range
     0.49%     0.49%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] xen_hypercall_xen_version
     0.47%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] dm_thin_find_mapped_range
     0.43%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] __find_block
     0.34%     0.34%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] __raw_callee_save___pv_queued_spin_unlock
     0.32%     0.24%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] mutex_lock
     0.27%     0.02%  kworker/u24:8-k  [dm_persistent_data]                              [k] bl_up_read
     0.23%     0.04%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] dm_bufio_release
     0.23%     0.00%  kworker/u24:8-k  [dm_persistent_data]                              [k] dm_btree_remove
     0.23%     0.00%  kworker/u24:8-k  [dm_persistent_data]                              [k] exit_shadow_spine
     0.22%     0.19%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] __relink_lru
     0.22%     0.22%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] mutex_unlock
     0.21%     0.21%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] _raw_spin_lock
     0.20%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] passdown_double_checking_shared_status
     0.19%     0.19%  kworker/u24:8-k  [dm_persistent_data]                              [k] __add_holder
     0.19%     0.00%  kworker/u24:8-k  [dm_thin_pool]                                    [k] dm_pool_block_is_shared
     0.19%     0.00%  kworker/u24:8-k  [dm_persistent_data]                              [k] sm_ll_lookup
     0.19%     0.00%  kworker/u24:8-k  [dm_persistent_data]                              [k] remove_raw
     0.18%     0.18%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] xen_restore_fl_direct
     0.18%     0.03%  kworker/u24:8-k  [dm_persistent_data]                              [k] bl_down_write
     0.17%     0.01%  kworker/u24:8-k  [dm_persistent_data]                              [k] exit_ro_spine
     0.16%     0.01%  kworker/u24:8-k  [dm_persistent_data]                              [k] bl_up_write
     0.15%     0.01%  Xorg             libc-2.24.so                                      [.] __GI___ioctl
     0.15%     0.15%  kworker/u24:8-k  [dm_persistent_data]                              [k] __del_holder
     0.14%     0.12%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] _cond_resched
     0.13%     0.07%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] __flush_write_list
     0.12%     0.00%  kworker/u24:10-  [kernel.kallsyms]                                 [k] ret_from_fork
     0.12%     0.00%  kworker/u24:10-  [kernel.kallsyms]                                 [k] kthread
     0.12%     0.00%  kworker/u24:10-  [kernel.kallsyms]                                 [k] worker_thread
     0.12%     0.00%  kworker/u24:10-  [kernel.kallsyms]                                 [k] process_one_work
     0.12%     0.00%  libvirtd         [unknown]                                         [.] 0000000000000000
     0.12%     0.00%  libvirtd         libc-2.24.so                                      [.] __GI___ioctl
     0.12%     0.12%  libvirtd         [kernel.kallsyms]                                 [k] xen_hypercall_domctl
     0.12%     0.00%  libvirtd         [unknown]                                         [k] 0x0000000000000016
     0.12%     0.12%  kworker/u24:8-k  [dm_persistent_data]                              [k] lower_bound
     0.11%     0.00%  Xorg             [kernel.kallsyms]                                 [k] entry_SYSCALL_64_after_hwframe
     0.11%     0.00%  Xorg             [kernel.kallsyms]                                 [k] do_syscall_64
     0.11%     0.00%  Xorg             [kernel.kallsyms]                                 [k] __x64_sys_ioctl
     0.11%     0.07%  kworker/u24:8-k  [kernel.kallsyms]                                 [k] dm_bufio_mark_partial_buffer_dirty
     0.11%     0.00%  Xorg             [nouveau]                                         [k] nouveau_drm_ioctl
     0.11%     0.11%  swapper          [kernel.kallsyms]                                 [k] xen_hypercall_xen_version
     0.11%     0.00%  Xorg             [drm]                                             [k] drm_ioctl
     0.10%     0.00%  Xorg             [drm]                                             [k] drm_ioctl_kernel
     0.10%     0.00%  kworker/u24:10-  [dm_thin_pool]                                    [k] do_worker
```

**Comment by DemiMarie** [comment]
> How about performing `blkdiscard` asynchronously? Possible solution would be to rename the LV to be removed to `vm-*-stale`, put it to reclaim queue, do blkdiscard, and finally run lvremove on it.

If this bug causes I/O on other devices in the pool to hang, I do not believe this would help.  Right now, the only options I know of are:

1. Fix the kernel bug.
2. Switch to a different storage pool.
3. Do not call `blkdiscard` before destroying an LVM thin volume, and rely on `thin_trim` instead.

**Comment by DemiMarie** [comment]
Setting up `thin_trim` will be tricky, as LVM2 does not support it.  A workaround exists but appears to be incredibly error-prone.

**Comment by brendanhoar** [comment]
> 1. Fix the kernel bug.
> 

Do we know the specifics of the particular kernel bug causing this?

Brendan

**Comment by DemiMarie** [comment]
Looks like the problem is a missing API in the kernel btree code: it needs to support native range deletions, instead of `dm-thin` having to (badly) emulate them.

**Comment by iamahuman** [comment]
> > How about performing `blkdiscard` asynchronously? Possible solution would be to rename the LV to be removed to `vm-*-stale`, put it to reclaim queue, do blkdiscard, and finally run lvremove on it.
> 
> If this bug causes I/O on other devices in the pool to hang, I do not believe this would help.

Sorry for the unclear comment. By "asynchronously" I actually meant a (hacky) workaround where the daemon splits the `blkdiscard` request into smaller ranges, so it wouldn't take up the I/O queue for a long time and instead let other BIOs pass through between successive discard requests.

It's not ideal, but it seems to work at the moment. I have used this approach for some time in a form of a custom extension to qubesd (this work was trivial, but a kernel patch would not be; I'm afraid I won't have enough time to write a proper kernel patch for now).

> Right now, the only options I know of are:
> 
>     1. Fix the kernel bug.
> 
>     2. Switch to a different storage pool.
> 
>     3. Do not call `blkdiscard` before destroying an LVM thin volume, and rely on `thin_trim` instead.

**Comment by iamahuman** [comment]
> To the people who experience these shutdown delays: If you shut down many VMs in parallel with `qvm-shutdown --wait vm1 vm2 vm3...`, is the total delay the same as the sum of the individual delays, less, or more?

Theoretically less in that the guest-side shutdown process is done in parallel, but the discard I/Os would still be serialized, resulting in even _longer_ freeze time.

**Comment by DemiMarie** [comment]
> > To the people who experience these shutdown delays: If you shut down many VMs in parallel with `qvm-shutdown --wait vm1 vm2 vm3...`, is the total delay the same as the sum of the individual delays, less, or more?
> 
> Theoretically less in that the guest-side shutdown process is done in parallel, but the discard I/Os would still be serialized, resulting in even _longer_ freeze time.

I think the idea is to insert breaks in the I/O stream, to give other I/O a chance to happen.

**Comment by brendanhoar** [comment]
Could this be as simple as modifying /usr/lib/python3.5/site-packages/qubes/storage/lvm.py and change:

pre_cmd = ['blkdiscard', '/dev/'+cmd[1]]

To:

pre_cmd = ['blkdiscard', '--step 1073741824','/dev/'+cmd[1]]

and reboot?

Might have to dial in the step size (that's 1GB).

**Comment by DemiMarie** [comment]
> It's not ideal, but it seems to work at the moment. I have used this approach for some time in a form of a custom extension to qubesd (this work was trivial, but a kernel patch would not be; I'm afraid I won't have enough time to write a proper kernel patch for now).

How large a change would the kernel patch be, and where in the kernel should I look?

**Comment by DemiMarie** [comment]
> One way that comes to my mind is to locate the "dominators" of all the leaf nodes in the range, and call `dm_btree_del` on them. This way, we can save much of re-balancing work while preserving refcount book-keeping.

Looks like the kernel supports range deletions, at least in git HEAD.  So the hard part is “just” hooking up `dm-thin` to call that code.

**Comment by iamahuman** [comment]
@DemiMarie Would you mind posting the link? For now I can't find anything other than the emulation (https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/tree/drivers/md/dm-thin-metadata.c).

**Comment by DemiMarie** [comment]
> @DemiMarie Would you mind posting the link? For now I can't find anything other than the emulation (https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/tree/drivers/md/dm-thin-metadata.c).

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/md/persistent-data/dm-btree-remove.c?id=dbb5afad100a828c97e012c6106566d99f041db6#n668

**Comment by ejose19** [comment]
> Could this be as simple as modifying /usr/lib/python3.5/site-packages/qubes/storage/lvm.py and change:
> 
> pre_cmd = ['blkdiscard', '/dev/'+cmd[1]]
> 
> To:
> 
> pre_cmd = ['blkdiscard', '--step 1073741824', '/dev/'+cmd[1]]
> 
> and reboot?
> 
> Might have to dial in the step size (that's 1GB). Also not sure if that's correct syntax.

This does alleviates the issue considerably, there's no full freeze like before although on some shutdowns (not on all for the same VM) I noticed a bit of lag, but it was tolerable. Interestingly though, shutdown time was also reduced (around ~10%, as measured by `time qvm-shutdown`), and VMs with small storage doesn't seem to be affected by this change.

Here's the full line I've used:
```python
pre_cmd = ['blkdiscard', '-p', '1G', '/dev/'+cmd[1]]
```

thanks @brendanhoar!

**Comment by brendanhoar** [comment]
> Here's the full line I've used:
> 
> ```python
> 
> pre_cmd = ['blkdiscard', '-p', '1G', '/dev/'+cmd[1]]
> 
> ```
> thanks @brendanhoar!

Awesome, glad to hear it worked!

:)

-B

**Comment by brendanhoar** [comment]
Created a pull request here:

https://github.com/QubesOS/qubes-core-admin/pull/404

**Comment by iamahuman** [comment]
> > @DemiMarie Would you mind posting the link? For now I can't find anything other than the emulation (https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/tree/drivers/md/dm-thin-metadata.c).
> 
> https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/md/persistent-data/dm-btree-remove.c?id=dbb5afad100a828c97e012c6106566d99f041db6#n668

I don't see any change here. `dm_btree_remove_leaves` calls into `remove_one`, which in turn uses `remove_nearest` to find nodes to delete. Note that `remove_nearest` can only find/return leaf nodes, which means that `remove_one` can only operate on one leaf node at a time.

Also the level part is a completely different thing; ii is designed to operate on 2+ levels of mappings with mutually different purposes (e.g. thin volume btree -> allocation map btree).

**Comment by DemiMarie** [comment]
> > > @DemiMarie Would you mind posting the link? For now I can't find anything other than the emulation (https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/tree/drivers/md/dm-thin-metadata.c).
> > 
> > 
> > https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/md/persistent-data/dm-btree-remove.c?id=dbb5afad100a828c97e012c6106566d99f041db6#n668
> 
> I don't see any change here. `dm_btree_remove_leaves` calls into `remove_one`, which in turn uses `remove_nearest` to find nodes to delete. Note that `remove_nearest` can only find/return leaf nodes, which means that `remove_one` can only operate on one leaf node at a time.

That’s unfortunate :disappointed:.  Would you be willing and able to guide me through fixing this properly, or is this something that is not a good start for a kernel newbie?  A proper fix would be greatly appreciated.

> Also the level part is a completely different thing; ii is designed to operate on 2+ levels of mappings with mutually different purposes (e.g. thin volume btree -> allocation map btree).

Good to know, thanks!

**Comment by heinrich-ulbricht** [comment]
Wow, lot's of activity here. 

Do I understand correctly that there are two issues that are being currently discussed:

1. shutdown for huge AppVMs on HDDs take lots of time; the bigger the AppVM the longer the shutdown
2. shutting down a huge VM may block the whole system for some time

And that the fix by @brendanhoar alleviates the second issue?

**Comment by DemiMarie** [comment]
> Wow, lot's of activity here.
> 
> Do I understand correctly that there are two issues currently being discussed:
> 
> 1. shutdown for huge AppVMs on HDDs take lots of time; the bigger the AppVM the longer the shutdown
> 2. shutting down a huge AppVM may block the whole system for some time
> 
> And that the fix by @brendanhoar alleviates the second issue?

It should fix the second issue, yes.  The first issue is blocked on @iamahuman having enough time to write a proper kernel patch.

**Comment by brendanhoar** [comment]
Will these (announced today) help with the thin-pool/shutdown performance...in particular, the first one?

https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/commit/?h=dm-5.14&id=c270679556fe2224bbbacd92fd9a4fad204654a9
https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/commit/?h=dm-5.14&id=aa6f12ee4302175540a052a9303c4e4127b3f267
https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/commit/?h=dm-5.14&id=bfa37c3fed7b66dabe0bf7eda532a0ce16367ca7

**Comment by marmarek** [comment]
Yes, sounds very likely. We'll see when 5.14 will be out.

**Comment by iamahuman** [comment]
> > Wow, lot's of activity here.
> > Do I understand correctly that there are two issues currently being discussed:
> > 
> > 1. shutdown for huge AppVMs on HDDs take lots of time; the bigger the AppVM the longer the shutdown
> > 2. shutting down a huge AppVM may block the whole system for some time
> > 
> > And that the fix by @brendanhoar alleviates the second issue?
> 
> It should fix the second issue, yes. The first issue is blocked on @iamahuman having enough time to write a proper kernel patch.

Since I don't believe I would be able to make enough time for this anytime soon, maybe I could at least put together a proposal and mail dm-devel? What should be in there, though?

Also I'm not exactly an active kernel contributor either.

**Comment by iamahuman** [comment]
> Will these (announced today) help with the thin-pool/shutdown performance...in particular, the first one?
> 
>     1. dm space maps: improve performance with inc/dec on ranges of blocks:
>        https://git.kernel.org/pub/scm/linux/kernel/git/device-mapper/linux-dm.git/commit/?h=dm-5.14&id=c270679556fe2224bbbacd92fd9a4fad204654a9

This is certainly in the right direction, but it still misses some of the most important pieces required to eliminate the root cause: range _deletion_ of B-tree nodes. Right now I'm split between taking the B-tree approach or tree consolidation (put together discarded blocks in some other tree for durability and start from there).

**Comment by iamahuman** [comment]
Not a particularly efficient or ideal approach, but I'm using an extension I wrote myself that 1. splits discard requests and 2. do them in the background so that it doesn't block qubesd from performing other tasks.

**Comment by DemiMarie** [comment]
> This is certainly in the right direction, but it still misses some of the most important pieces required to eliminate the root cause: range _deletion_ of B-tree nodes. Right now I'm split between taking the BTRFS approach and performing tree consolidation (put together discarded blocks in some other tree for durability and start from there).

What is the BTRFS approach?  We already support BTRFS storage pools, so I am a bit confused.

**Comment by DemiMarie** [comment]
> Since I don't believe I would be able to make enough time for this anytime soon, maybe I could at least put together a proposal and mail dm-devel? What should be in there, though?

I honestly have no idea.  You at least seem to understand this code :smile:.

> Also I'm not exactly an active kernel contributor either.

At least you understand the relevant kernel code, which I do not believe any of the Qubes core team does.

**Comment by iamahuman** [comment]
> > Not a particularly efficient or ideal approach, but I'm using an extension I wrote myself that 1. splits discard requests and 2. do them in the background so that it doesn't block qubesd from performing other tasks.
> 
> Do you clone the volume, then after removal of the original, pass the clone to a thread that does throttled discards (and an lvremove of the clone after they complete)? Or do you just rename the volume, hand the renamed volume to a thread that does the throttled blkdiscard and lvremove, and allow the qubes storage drive to think it was removed?

It's the latter. To be accurate, the Qubes storage driver already forgets the volume by the time it is renamed (as if it was deleted, which means it uses `lvrename` just like `lvremove`), and the knowledge of the volume's existence is handed over exclusively to the background thread. As for how it is renamed, qubesd would prepend a `RECLAIM-` prefix to the volumes put into the removal queue, so that qubesd could scan for those volumes next time it restarts. This way qubesd can resume the "throttled discard" operation if it gets interrupted due to e.g. power failure.

**Comment by DemiMarie** [comment]
> > > Not a particularly efficient or ideal approach, but I'm using an extension I wrote myself that 1. splits discard requests and 2. do them in the background so that it doesn't block qubesd from performing other tasks.
> > 
> > 
> > Do you clone the volume, then after removal of the original, pass the clone to a thread that does throttled discards (and an lvremove of the clone after they complete)? Or do you just rename the volume, hand the renamed volume to a thread that does the throttled blkdiscard and lvremove, and allow the qubes storage drive to think it was removed?
> 
> It's the latter. To be accurate, the Qubes storage driver already forgets the volume by the time it is renamed (as if it was deleted, which means it uses `lvrename` just like `lvremove`), and the knowledge of the volume's existence is handed over exclusively to the background thread. As for how it is renamed, qubesd would prepend a `RECLAIM-` prefix to the volumes put into the removal queue, so that qubesd could scan for those volumes next time it restarts. This way qubesd can resume the "throttled discard" operation if it gets interrupted due to e.g. power failure.

Naming VM volumes anything that does not start with `vm-` is a bad idea unless the relevant udev rules are changed: they could get scanned by udev.

**Comment by iamahuman** [comment]
> > > > Not a particularly efficient or ideal approach, but I'm using an extension I wrote myself that 1. splits discard requests and 2. do them in the background so that it doesn't block qubesd from performing other tasks.
> > > 
> > > 
> > > Do you clone the volume, then after removal of the original, pass the clone to a thread that does throttled discards (and an lvremove of the clone after they complete)? Or do you just rename the volume, hand the renamed volume to a thread that does the throttled blkdiscard and lvremove, and allow the qubes storage drive to think it was removed?
> > 
> > 
> > It's the latter. To be accurate, the Qubes storage driver already forgets the volume by the time it is renamed (as if it was deleted, which means it uses `lvrename` just like `lvremove`), and the knowledge of the volume's existence is handed over exclusively to the background thread. As for how it is renamed, qubesd would prepend a `RECLAIM-` prefix to the volumes put into the removal queue, so that qubesd could scan for those volumes next time it restarts. This way qubesd can resume the "throttled discard" operation if it gets interrupted due to e.g. power failure.
> 
> Naming VM volumes anything that does not start with `vm-` is a bad idea unless the relevant udev rules are changed: they could get scanned by udev.

Thanks! Forgot to mention it...

**Comment by faerbersteve** [comment]
Has someone tried using a newer version of libvirt?
The newer version 7.1 of libvirt contains a fix for a shutdown race which is mentioned in the release notes.
Maybe this solve the issue or part of it.

**Comment by marmarek** [comment]
On Thu, Jun 03, 2021 at 10:35:31AM -0700, faerbersteve wrote:
> Has someone tried using a newer version of libvirt?
> The newer version 7.1 of libvirt contains a fix for a shutdown race which is mentioned in the release notes.
> Maybe this solve the issue or part of it.

That sounds like a very different issue. The problem we face here is
happening outside of libvirt.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab

**Comment by DemiMarie** [comment]
> Not a particularly efficient or ideal approach, but I'm using an extension I wrote myself that 1. splits discard requests and 2. do them in the background so that it doesn't block qubesd from performing other tasks.

Would you mind submitting a PR for this?  This seems like the best approach until a proper kernel patch can be written.  An email to dm-devel would also be great.

**Comment by iamahuman** [comment]
> > Not a particularly efficient or ideal approach, but I'm using an extension I wrote myself that 1. splits discard requests and 2. do them in the background so that it doesn't block qubesd from performing other tasks.
> 
> Would you mind submitting a PR for this? This seems like the best approach until a proper kernel patch can be written. An email to dm-devel would also be great.

The code is written for R4.0, so perhaps I'll be setting up a testing environment in the coming week. Might take a while...

**Comment by iamahuman** [comment]
It appears like I might not be able to procure enough time anytime soon, so here's the plugin code I had been using at least:

https://gist.github.com/iamahuman/a784efbbcadaec508235a10b52cc8baa

Sub-tasks for morphing it into something actually merge-able:

- [ ] Get rid of `linux_fs.py`, switch to `blkdiscard -p`
- [ ] Replace the worker pool with a purely asynchronous version
- [ ] Figure out appropriate lifecycle management scheme for the worker pool
  - [ ] Starting: `__init__` or something more appropriate
  - [ ] Stopping: the operation itself is durable, so we don't need to handle abrupt shutdown cases; we merely have to clean up pending `blkdiscard` processes on `systemctl stop qubesd`.
- [ ] Perhaps just make the queue simply local to `ThinPool`

**Comment by DemiMarie** [comment]
> * [ ]  Get rid of `linux_fs.py`, switch to `blkdiscard -p`

Is this actually necessary?  Using an ioctl will always be more efficient than spawning a process.

> * [ ]  `s/yield from/await/`

I would start with this, as it has no dependencies on any of the other changes.

**Comment by iamahuman** [comment]
> > * [ ]   Get rid of `linux_fs.py`, switch to `blkdiscard -p`
> 
> Is this actually necessary? Using an ioctl will always be more efficient than spawning a process.

Mainly to reduce complexity. However if we keep it, I suppose we shall keep the worker pool as well.

> 
> > * [ ]   `s/yield from/await/`
> 
> I would start with this, as it has no dependencies on any of the other changes.

**Comment by DemiMarie** [comment]
> > > * [ ]    Get rid of `linux_fs.py`, switch to `blkdiscard -p`
> > 
> > 
> > Is this actually necessary? Using an ioctl will always be more efficient than spawning a process.
> 
> Mainly to reduce complexity. However if we keep it, I suppose we shall keep the worker pool as well.

I will let @marmarek decide if the complexity is worthwhile.  On bleeding-edge kernels, we might be able to get rid of the pool with `io_uring`, but there is no way whatsoever that _that_ is worth the effort.

**Comment by q4747** [comment]
I can confirm that re-installing QubesOS 4.0.4 with BTRFS worked for me to solve this issue.

Previously I had QubesOS 4.0.4 with the default LVM-thin. This led to rather slow shutdowns and sometimes temporary GUI freezes when shutting down VMs with data of ~300GB and larger. It typically took 30 sec. up to 2 min, depending on the size.
Just like @heinrich-ulbricht, I was able to observe the progress of the slow deletion of the backup file with `sudo lvs | grep [vm-name]`.

I'm using the SSD Samsung 870 EVO with 4TB, which works fine now with BTRFS.

**Comment by DemiMarie** [comment]
I think using `thin_trim` during boot is definitely a good idea.  In the long term this needs to be solved at the kernel level.

**Comment by MonkeyBars3k** [comment]
Confirming that fresh install of 4.1 with BTRFS enabled solves this issue.

The instructions on how to select this option during the install setup [are detailed here](https://forum.qubes-os.org/t/how-does-the-vm-storage-work-internally/5355/5).

**Comment by rustybird** [comment]
Even if LVM currently doesn't have a dedicated defrag command, wouldn't it still be possible to defragment a volume in a generic way:

0. Ensure that there's plenty of free space in the pool
1. Wait for the problematic VM to cleanly shut down, however long it takes
2. Delete the fragmented `save_on_stop` volume's revisions (except the latest one!)
3. Create a new thin volume of the same size
4. `cp --sparse=always` from the old volume's to the new volume's `/dev` node
5. Delete the old volume
6. Rename the new volume to the old name
7. Restart `qubesd.service`

This is totally untested and maybe I'm missing something due to lack of familiarity with LVM2 and the lvm_thin driver.

**Comment by johnnyboy-3** [comment]
Hello,

with R4.2 things have changed.
With default partition scheme this problem no longer exists.
Interestingly this problem now occurs **with** BTRFS (both tested with upgrade and fresh installation).
I do like the compress feature that BTRFS supports, is there a way to enable it with lvm/ext4?
Or is there a fix in the near future, that solves this issue with BTRFS?

See also small discussion @ https://forum.qubes-os.org/t/very-slow-shutdown-of-large-vms-in-4-2-even-with-btrfs/23156
